js/components/common/react_modal.js.jsx

```diff
@@ -34,19 +34,86 @@ class CourseModal extends React.Component {
       course: [],
       sessions: [],
       courseTitle: "",
+      clicked: 0,
     }
   }
+  /**
+   * Change the clicked and courseId state, whenever a course link is clicked.
+   */
+  clickRelated = s => {
+    let copyState = this.state
+    copyState.courseId = s
+    copyState.clicked = 1
+    this.setState(copyState)
+  }
+
+  /**
+   * Convert the course names to <a> tags and return a list of strings and <a> tags.
+   */
+  getRelatedCourses = content => {
+    const result = []
+    if (content !== null) {
+      content.split(" ").forEach((word, i) => {
+        if (word.match(/[A-Z]{3,4}[0-9]{3}[H|Y]1/gi)) {
+          let symbol = ""
+          // check if the last character is not 1 (is a symbol). If so, remove it from word and
+          // add it as a separate string to the list.
+          const sessions = ["H1", "Y1"]
+          sessions.forEach(session => {
+            if (
+              word.indexOf(session) !== word.length - 1 &&
+              word.indexOf(session) !== -1
+            ) {
+              symbol = word.substring(word.indexOf(session) + 2)
+              word = word.substring(0, word.indexOf(session) + 2)
+            }
+          })
+
+          result.push(
+            <a
+              key={i}
+              className="course-selection"
+              onClick={() => this.clickRelated(word)}
+            >
+              {word}
+            </a>
+          )
+          result.push(symbol)
+        } else {
+          result.push(word)
+        }
+        result.push(" ")
+      })
+    }
+    return result
+  }
 
   componentDidUpdate(prevProps) {
-    if (prevProps.courseId !== this.props.courseId && this.props.courseId !== "") {
-      getCourse(this.props.courseId).then(course => {
+    let changed = 0
+    if (this.props.showCourseModal !== prevProps.showCourseModal) {
```

Comment:
Overall the management of the `courseId` has gotten more complex, since we need to take into account both the props (when the user opens a modal from the graph) and changes to the state (when a user clicks on the link in the existing modal).

To keep the logic of `componentDidUpdate` relatively simple, and also preserve the caching behaviour for when a user opens the modal from the same course twice, please make the following changes:

1. If the prop `courseId` changes (which is what the original code checked), don't call `getCourse`, instead `setState` to modify the `state.courseId`. (This will trigger another call to `componentDidUpdate`.)
2. If the *state* `courseId` changes, call `getCourse`. This should then handle both types of "course id state changes", and remove the need for the `clicked` state that you added.

Reply:
Right, what I mean by "caching" here is that if the user opens the modal for the same course twice, the data is fetched only once (`getCourse`), precisely because `prevProps.courseId === this.props.courseId`.